apk/apk.go

```diff
@@ -0,0 +1,424 @@
+
+	err := tw.WriteHeader(header)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func writeFile(tw *tar.Writer, header *tar.Header, file io.Reader) error {
+	// if hash {
+	// 	header.Format = tar.FormatPAX
+
+	// 	digest := sha1.New()
+	// 	_, err := io.Copy(digest, file)
+	// 	if err != nil {
+	// 		return err
+	// 	}
+
+	// 	header.PAXRecords = map[string]string{
+	// 		"APK-TOOLS.checksum.SHA1": "f572d396fae9206628714fb2ce00f72e94f2258f",
+	// 		// "APK-TOOLS.checksum.SHA1": hex.EncodeToString(digest.Sum(nil)),
+	// 	}
+	// } else {
+	header.Format = tar.FormatUSTAR
+	header.ChangeTime = time.Time{}
+	header.AccessTime = time.Time{}
+	// }
+
+	err := tw.WriteHeader(header)
+	if err != nil {
+		return err
+	}
+
+	_, err = io.Copy(tw, file)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type tarKind int
+
+const (
+	tarFull tarKind = iota
+	tarCut
+)
+
+func writeTgz(w io.Writer, kind tarKind, builder func(tw *tar.Writer) error, digest hash.Hash) ([]byte, error) {
+	mw := io.MultiWriter(digest, w)
+	gw := gzip.NewWriter(mw)
+	cw := newWriterCounter(gw)
+	bw := bufio.NewWriterSize(cw, 4096)
+	tw := tar.NewWriter(bw)
+
+	err := builder(tw)
+	if err != nil {
+		return nil, err
+	}
+
+	// handle the cut vs full tars
+	if err = bw.Flush(); err != nil {
+		return nil, err
+	}
+	if err = tw.Close(); err != nil {
+		return nil, err
+	}
+	if kind == tarFull {
+		if err = bw.Flush(); err != nil {
+			return nil, err
+		}
+	}
+
+	size := cw.Count()
+	alignedSize := (size + 511) & ^uint64(511)
+
+	increase := alignedSize - size
+	if increase > 0 {
+		b := make([]byte, increase)
+		_, err = cw.Write(b)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if err = gw.Close(); err != nil {
+		return nil, err
+	}
+
+	return digest.Sum(nil), nil
+}
+
+func parseRsaPrivateKeyFromPemStr(privPEM string) (*rsa.PrivateKey, error) {
+	block, _ := pem.Decode([]byte(privPEM))
+	if block == nil {
+		return nil, errors.New("failed to parse PEM block containing the key")
+	}
+
+	priv, err := x509.ParsePKCS1PrivateKey(block.Bytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return priv, nil
+}
+
+/*func main() {
+	if err := runit("foo", "../alpine/user.rsa", "", os.Args[1]); err != nil {
+		log.Fatalln(err)
+	}
+}
+*/
+
+func runit(pathToFiles, pathToKey, workDir, target string) (err error) {
+	signatureTgz, err := os.Create(path.Join(workDir, "apk_signatures.tgz"))
+	if err != nil {
+		return err
+	}
+	defer signatureTgz.Close()
+
+	controlTgz, err := os.Create(path.Join(workDir, "apk_control.tgz"))
+	if err != nil {
+		return err
+	}
+	defer controlTgz.Close()
+
+	dataTgz, err := os.Create(path.Join(workDir, "apk_data.tgz"))
+	if err != nil {
+		return err
+	}
+	defer dataTgz.Close()
+
+	size := int64(0)
+	sizep := &size
+
+	// create the data tgz
+	dataDigest, err := createGzData(dataTgz, pathToFiles, sizep, size)
+	if err != nil {
+		return err
+	}
+
+	// create the control tgz
+	builderControl := createBuilderControl(size, dataDigest, err)
```

Comment:
Maybe use the same pattern as for the `data` section using `createControl` function?